src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -1072,6 +1103,150 @@ void DatabaseCatalog::updateLoadingDependencies(const StorageID & table_id, Tabl
     old_dependencies = std::move(new_dependencies);
 }
 
+void DatabaseCatalog::cleanupStoreDirectoryTask()
+{
+    fs::path store_path = fs::path(getContext()->getPath()) / "store";
+    size_t affected_dirs = 0;
+    for (const auto & prefix_dir : fs::directory_iterator{store_path})
+    {
+        String prefix = prefix_dir.path().filename();
+        bool expected_prefix_dir = prefix_dir.is_directory() &&
+            prefix.size() == 3 &&
+            isHexDigit(prefix[0]) &&
+            isHexDigit(prefix[1]) &&
+            isHexDigit(prefix[2]);
+
+        if (!expected_prefix_dir)
+        {
+            LOG_WARNING(log, "Found invalid directory {}, will try to remove it", prefix_dir.path().string());
```

Comment:
"invalid" --> "stale"?

Reply:
In this case directory is rather invalid than stale, because if has unexpected name. Such directory may appear in two cases: due to a bug in ClickHouse or due to manual intervention.